Handle federated service name lookups in kube-dns.

madhusudancs opened the pull request:
For the domain name queries that fail to match any records in the local
kube-dns cache, we check if the queried name matches the federation
pattern. If it does, we send a CNAME response to the federated name.

For more details look at the comments in the code.

Tests are coming ...

Also, this PR is based on @ArtfulCoder's PR #23930. So please review only the last commit here.

PTAL @ArtfulCoder @thockin @quinton-hoole @nikhiljindal 

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

madhusudancs committed:
Handle federated service name lookups in kube-dns.

For the domain name queries that fail to match any records in the local
kube-dns cache, we check if the queried name matches the federation
pattern. If it does, we send a CNAME response to the federated name.

For more details look at the comments in the code.

quinton-hoole commented on cmd/kube-dns/app/options/options.go:
```diff
@@ -33,6 +33,7 @@ type KubeDNSConfig struct {
 	KubeConfigFile string
 	KubeMasterURL  string
 	HealthzPort    int
+	Federations    map[string]string
```
Need a comment to explain what the key and value are here.

madhusudancs replied:
Added a short comment. Please let me know if you think it is sufficient.

quinton-hoole commented on cmd/kube-dns/app/options/options.go:
```diff
@@ -95,9 +97,44 @@ func (m kubeMasterURLVar) Type() string {
+	nameDomainMap map[string]string
+}
+
+func (fv federationsVar) Set(v string) error {
```
Better argument names please.  'federations' and 'keyAndVal' perhaps.  Please explain the format of this parameter in a comment.

quinton-hoole replied:
Aah - I see it's described below.  Perhaps a reference to that description here will help.

madhusudancs replied:
Added a comment/description here which is more relevant to this method. 

Changed the method argument to `keyVal`. It is a Go convention to use short names for method receivers, mostly 1-2 characters long. Also, all other receivers have a single letter names. So leaving this as is.

quinton-hoole commented on cmd/kube-dns/app/options/options.go:
```diff
@@ -95,9 +97,44 @@ func (m kubeMasterURLVar) Type() string {
+	return strings.Join(splits, ",")
+}
+
+func (fv federationsVar) Type() string {
```
This doesn't seem to be called? What is it for?

madhusudancs replied:
This is to satisfy the pflag's `Value` interface so that we can pass it as an argument to fs.Var here - https://github.com/kubernetes/kubernetes/pull/25727/commits/3ad8ad1cd287055947d5a0669b3d1a61f19cbbd6#diff-0f7323379621678150800d4b5da58230R139

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -48,28 +49,38 @@ const (
 	resyncPeriod = 30 * time.Minute
+
+	// Duration for which the TTL cache should hold the node.
+	nodeCacheTTL = 180 * time.Second
```
Need to explain why this value was chosen please. What are the pros and cons of increasing or decreasing this value. Why is there a need to evict cache entries at all?  Are they not automatically kept up to date via watches?

madhusudancs replied:
Added a clarification comment. We are not setting a watch here. We just need one node and our assumption is that the zone name annotation on the nodes don't change often, so there is not really a need to observe and react to the nodes' state changes. But we don't want to hold the values for ever without letting it ever change. So the value is chosen to be neither too long nor too short.

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -48,28 +49,38 @@ const (
 	domain string
+	// Map of federation names that the cluster in which this kube-dns is running belongs to, to
+	// the corresponding domain names.
+	federations map[string]string
```
Isn't this supposed to be type federationsVar?

madhusudancs replied:
No. There is a little subtlety here. `federationsVar` is a struct that satisfies the `Value` interface and wraps around the `nameDomainMap` map in which we are really interested. We ultimately want to have the command line flag values in the `federations` map here because that's all we care about, the map. So we alias the `federationsVar.nameDomainMap` to point to `federations`, so when pflag deserializes the input command line values by calling the `federationVars.Set()` method, the key-values are copied to `federationsVar.nameDomainMap` and hence `federations` map here.

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -48,28 +49,38 @@ const (
 	endpointsStore kcache.Store
 	// A cache that contains all the services in the system.
-	servicesStore     kcache.Store
+	servicesStore kcache.Store
```
Formatting got messed up.  I don't think that this line needs to change at all?

madhusudancs replied:
I don't think it got messed up. It's just the way gofmt works. It is the result of adding a comment line after `serviceStore kcache.Store` line. Before the comment line, gofmt tried to align the type with the types in the following lines. But after inserting a comment it lost that constraint.

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -48,28 +49,38 @@ const (
-	servicesStore     kcache.Store
+	servicesStore kcache.Store
+	// A TTL cache that contains some subset of nodes in the system.
+	nodesStore        kcache.Store
```
I'm not yet clear why wee need this.  Hopefully I'll find out soon.

madhusudancs replied:
Added a comment to clarify this a bit. Please let me know if that is sufficient.

madhusudancs committed:
Switch kube-dns to use external versioned API instead of the internal version.

madhusudancs committed:
To get the zone information, select a node which actually has that annotation.

Also, use the cluster zone information while generating the CNAME
response for federation queries.

madhusudancs committed:
Implement tests for federation queries.

madhusudancs commented:
@quinton-hoole thanks for the review. I will address your comments. In the meantime, I have pushed additional commits that implement tests and include some code refactor. PTAL.

@ArtfulCoder @thockin I think you should take a look too.

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -338,10 +349,17 @@ func (kd *KubeDNS) Records(name string, exact bool) ([]skymsg.Service, error) {
+
+	// If the name query is not an exact query and does not match any records in the local store,
+	// attempt to send a federation redirect (CNAME) response.
+	if !exact {
```
I don't understand what an 'exact' query is, and why you would not want to handle it like any other request.  Can you add a comment to explain?

madhusudancs replied:
It's not very well documented, but if I understand it correctly, it means that SkyDNS wants the query names to be interpreted as is, without any modifications. Since we are going to modify the query name and send it as a CNAME response I think it is not appropriate to do it when it is true.

I could be wrong here.

@ArtfulCoder @thockin please correct me if I am wrong.

madhusudancs commented:
@quinton-hoole addressed the comments, PTAL.

quinton-hoole commented:
@madhusudancs LGTM.  This will need a rebase now that #23930 is in.  Adding LGTM in the mean time.  Feel free to re-add yourself once the rebase against #23930 is in.

quinton-hoole commented:
And you also have a godep license issue.  Best fix that up during the rebase too...

```
FAILED   ./hack/../hack/verify-godep-licenses.sh	16s
```
If I'm reading the error log correctly, it's caused by this:

```
= vendor/go4.org/errorutil licensed under: =
 
-<html>
-<head>
-<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
-<meta name="go-import" content="go4.org git https://github.com/camlistore/go4">
-<meta name="go-source" content="go4.org https://github.com/camlistore/go4/ https://github.com/camlistore/go4/tree/master{/dir} https://github.com/camlistore/go4/blob/master{/dir}/{file}#L{line}">
-
-<meta http-equiv="refresh" content="0; url=https://godoc.org/go4.org/errorutil/master/LICENSE.md">
-</head>
-<body>
-See <a href="https://godoc.org/go4.org/errorutil/master/LICENSE.md">docs</a>.
-</body>
-</html>
-
+UNKNOWN
```

madhusudancs commented:
> And you also have a godep license issue. Best fix that up during the rebase too...

> ```
> FAILED   ./hack/../hack/verify-godep-licenses.sh    16s
> ```

> If I'm reading the error log correctly, it's caused by this:

I think this wasn't in my code because I did not touch anything in the vendor directory. I guess that was coming from the base "one container ..." PR. I have rebased to master now, so hopefully this error is gone.

madhusudancs commented:
Everything looks good, adding LGTM back.
